File: web_version/pmugg_files/guidelines/guideMutator.js
ms.guideMutator = function(classifier, timer, notifyFunc) {
	this.classifier = classifier;
	this.timer = timer;
	this.notify = notifyFunc;
	this.tasks = [];
	this.status = ms.guideMutator.status.SUCCESS;
	this.breakTime = 0;
	this.earlyEndTime = 0;
	this.endTime = 0;
	this.seedCount = 0;
	this.pauseCount = -1;
	this.nodeStats = new ms.nodeStats();
	this.optimizer = new ms.optimizer(this.nodeStats);
	
	this.hierarchyMutator = ms.globalSettings.get('Use Network') ?
		new ms.networkMutator(classifier, this.nodeStats) :
		new ms.familyTreeMutator(classifier, this.nodeStats);

	// This is only for debugging. We include this as a watch statement so it gets redrawn.
	ms.globalMutator = this;
	ms.guideMutator.forceContinued = false;
};

ms.guideMutator.taskCount = 0;
ms.guideMutator.findMutableVertexAttempts = 20;
ms.guideMutator.findMutableLineAttempts = 20;
ms.guideMutator.forceContinued = false;

ms.guideMutator.prototype.seed = function(mutationArea) {
	this.hierarchyMutator.setMutationArea(mutationArea);
	this.seedCount = 0;
};

ms.guideMutator.forceContinue = function() {
	ms.guideMutator.forceContinued = true;
};

ms.guideMutator.closeInspection = function(taskCount) {
	if (ms.guideMutator.forceContinued) {
		return false;
	}
	var taskStop = ms.globalSettings.get('Task Stop');
	var taskStep = ms.globalSettings.get('Task Step');
	if (taskStop >= 0) {
		return taskCount >= taskStop && ((taskCount - taskStop) % taskStep == 0);
	} else {
		return ms.globalSettings.get('Debug Each Task');
	}
};

ms.guideMutator.prototype.mutate = function() {
	var probabilities = [1, 1, 10];
	var done = false;
	var self = this;
	while (!done) {
		switch(ms.randomDistribution(probabilities)) {
			case 0:
				ms.timerG.start('Add Fragment');
				var success = this.hierarchyMutator.addStartInstance();
				ms.timerG.stop('Add Fragment');
				if (success) {
					return;
				} else {
					probabilities[0] = 0;
				}

				break;
			case 1:
				ms.timerG.start('Remove Fragment');
				var success = this.hierarchyMutator.changeRandomInstance(true);
				ms.timerG.stop('Remove Fragment');
				if (success) {
					return;
				} else {
					probabilities[1] = 0;
				}
				break;
			case 2:
				// TODO: Do not allow fully connected edges to be destroyed.
				ms.timerG.start('Modify Fragment');
				var success = this.hierarchyMutator.changeRandomInstance(false);
				ms.timerG.stop('Modify Fragment');
				if (success) {
					return;
				} else {
					probabilities[2] = 0;
				}
				break;
		}
		// Uncomment this part to force there to be fewer starter transitions.
		// We cannot skip a starter transition on the first task.
		if (ms.globalSettings.get('Fewer Start Transitions') &&
			(ms.guideMutator.taskCount > 1)) {
			return;
		}

		// The change has already been rejected.
		if (this.nodeStats.costChange['reject'] > 0) {
			return;
		}

		done = true;
		for (var i = 0; i < probabilities.length; i++) {
			if (probabilities[i] > 0) {
				done = false;
			}
		}
	}
};

ms.guideMutator.prototype.resolve = function(outputModel) {
	if (this.status == ms.guideMutator.status.PAUSED) {
		return this.status;
	}
	ms.timerG.start('Guide Mutator');
	while (true) {
		var taskCount = ++ms.guideMutator.taskCount;
		if (ms.guideMutator.closeInspection(taskCount)) {
			this.displayStats();
			debugger;
		}
		this.mutate();

		var cost = this.optimizer.computeCost();
		ms.timerG.start('Accept Or Reject');
		if (this.optimizer.isAccepted(cost)) {
			this.accept();
		} else {
			this.reject();
		}
		ms.timerG.stop('Accept Or Reject');
		ms.timerG.start('Save');
		this.nodeStats.save();
		ms.timerG.stop('Save');

		var time = Date.now();
		var maxTimeEnabled = ms.globalSettings.get('Max Time Enabled');
		var maxIterations = ms.globalSettings.get('Max Iterations');
		if ((maxTimeEnabled && (time > this.endTime || (time > this.earlyEndTime) && (this.nodeStats.getBadVertices().length == 0))) ||
			(taskCount >= maxIterations)){
			ms.timerG.stop('Guide Mutator');
			this.status = ms.guideMutator.status.FINISHED;
			return this.status;
		}
		if (taskCount == this.pauseCount) {
			this.pause();
			return this.status;
		}
		if (time > this.breakTime) {
			ms.timerG.stop('Guide Mutator');
			return ms.guideMutator.status.UNRESOLVED;
		}
	}
};

ms.guideMutator.prototype.accept = function() {
	if (this.nodeStats.getBadVertices().length == 0) {		
		this.seedCount++;
	}
	var taskCount = ms.guideMutator.taskCount;
	if (ms.guideMutator.closeInspection(taskCount)) {
		window.console.log(taskCount + ' accepted.');
	}
};

ms.guideMutator.prototype.reject = function() {
	var taskCount = ms.guideMutator.taskCount;
	this.nodeStats.restore();
	if (ms.guideMutator.closeInspection(taskCount) || ms.globalSettings.get('Debug Alerts')) {
		this.optimizer.verifyCost();
	}
	if (ms.guideMutator.closeInspection(taskCount)) {
		window.console.log(taskCount + ' rejected.');
	}
};

ms.guideMutator.prototype.findMutableVertex = function() {
	var vertexNodes = Object.values(this.nodeStats.nodes.vertex);
	if (vertexNodes.length > 0) {
		for (var i = 0; i < ms.guideMutator.findMutableVertexAttempts; i++) {
			var vertex = ms.pick(vertexNodes).getElement();
			if (vertex.isMoveable()) {
				return vertex;
			}
		}
	}
	return null;
};

ms.guideMutator.prototype.save = function(model) {
	var cost = this.optimizer.computeCost();
	this.optimizer.accept(cost);
	this.nodeStats.save();
};
	
ms.guideMutator.prototype.initializeModel = function(model) {
	this.edgeTypeStarted = [];
	this.pauseCount = -1;
	model.setNodeStats(this.nodeStats);

	var boundaryGroups = this.classifier.getBoundaryGroups();
	if (boundaryGroups && boundaryGroups.length > 0) {
		var ringInstance = new ms.ringInstance(this.classifier.getEmptyRing(), this.nodeStats);
		var ringGroup = ms.ringGroup.createFromRing(ringInstance.getRing());
		var startInstance = new ms.ringGroupInstance(ringGroup);
		startInstance.addRingInstance(0, ringInstance);
		var transition = {startInstance: startInstance, endGroup: ms.pick(boundaryGroups), initialBoundary: true};
		return this.hierarchyMutator.applyTransition(transition);
	} else {
		return true;
	}
};

ms.guideMutator.status = {
	FAILED: 0,
	UNRESOLVED: 1,
	PAUSED: 2,
	SUCCESS: 3,
	FINISHED: 4,
};

ms.guideMutator.prototype.getStatus = function() {
	return this.status;
};

ms.guideMutator.prototype.pause = function() {	
	this.displayStats();
	this.notify(true);
	this.status = ms.guideMutator.status.PAUSED;
};

ms.guideMutator.prototype.unresolve = function() {
	this.status = ms.guideMutator.status.UNRESOLVED;
};

ms.guideMutator.prototype.step = function(numTasks) {
	this.displayStats();
	this.pauseCount = ms.guideMutator.taskCount + numTasks;
};

ms.guideMutator.prototype.setBreakTime = function(breakTime) {
	this.breakTime = breakTime;
};

ms.guideMutator.prototype.setEndTime = function(earlyEndTime, endTime) {
	this.earlyEndTime = earlyEndTime;
	this.endTime = endTime;
};

ms.guideMutator.prototype.reset = function() {
	this.status = ms.guideMutator.status.UNRESOLVED;
};

ms.guideMutator.hideStats = function() {
	document.getElementById('statistics').innerHTML = '';
};

ms.guideMutator.prototype.displayStats = function() {
	var taskCount = ms.guideMutator.taskCount;
	var stats = 'Iteration: ' + taskCount + ' / ' + ms.globalSettings.get('Max Iterations');
	if (!ms.mvp) {
		stats += ' ' + this.optimizer.getPrevCost().sum.toFixed(1);
	}
	document.getElementById('statistics').innerHTML = stats;
};

ms.guideMutator.prototype.report = function() {
	var cost = this.optimizer.getPrevCost();
	var summary = cost.sum + ' ' + (this.timer.getTiming('everything') / 1000) + 's ';
	summary += (ms.guideMutator.taskCount / this.timer.getTiming('everything')).toFixed(4);
	var details = JSON.stringify(cost) + '\n';
	details += this.timer.reportMean();
	console.log('# of vertices: ' + this.nodeStats.getCount('vertex'));
	return {summary: summary, details: details};
};

File: cpp_version/pmugg/guidelines/guide_mutator.cpp
#include "pch.h"
#include "guide_mutator.h"
// #include "classifier.h"
// #include "timer.h"
// #include "model.h"
// #include "mutation_area.h"
// #include "hierarchy_mutator.h"
// #include "network_mutator.h"
// #include "family_tree_mutator.h"
#include "../grid/settings.h"
#include "../util/util.h"
#include <iostream>
#include <chrono>

namespace ms {

// bool GuideMutator::forceContinued = false;

GuideMutator::GuideMutator(Model* model, NetworkMutator* networkMutator)
    : // classifier(classifier)
    // , notify(notifyFunc)
    // , tasks()
    //, status(Status::SUCCESS)
    //, breakTime(0)
    //, earlyEndTime(0)
    //, endTime(0)
    //, seedCount(0)
    //, pauseCount(-1)
    // , nodeStats()
    // , optimizer(&nodeStats)
    model(model)
    , networkMutator(networkMutator) {
    
    // initializeHierarchyMutator();
    // ms::globalMutator = this;  // For debugging
}

//void GuideMutator::seed(MutationArea* mutationArea) {
//    hierarchyMutator->setMutationArea(mutationArea);
//    seedCount = 0;
//}

void GuideMutator::iterate(int steps) {
    timer->start("Guide Mutator");

    int finishStep = model->numSteps + steps;
    while (model->numSteps < finishStep) {
        std::cout << model->numSteps << " " << model->getCurrent()->getFaceMap().size() << std::endl;

        if (model->numSteps == 0) {
            mutateGround();
        } else {
            mutate();
        }

        // auto cost = optimizer.computeCost();
        timer->start("Accept Or Reject");
        // if (optimizer.isAccepted(cost)) {
        if (true) {
            accept();
        } else {
            reject();
        }
        timer->stop("Accept Or Reject");

        timer->start("Save");
        timer->stop("Save");

        auto now = std::chrono::system_clock::now();
        auto currentTime = std::chrono::duration_cast<std::chrono::milliseconds>(
            now.time_since_epoch()).count();

        bool maxTimeEnabled = globalSettings["Max Time Enabled"].get<bool>();
        int maxIterations = globalSettings["Max Iterations"].get<int>();

        // model->getCurrent()->save(std::to_string(model->numSteps));
        model->numSteps++;
    }
    timer->stop("Guide Mutator");
    // model->getCurrent()->save("");
}

// Add the ground plane.
void GuideMutator::mutateGround() {
    bool success = networkMutator->addStartInstance(true);
};

void GuideMutator::mutate() {
    std::vector<double> probabilities = {1, 1, 10};
    bool done = false;

    while (!done) {
        switch(Util::randomDistribution(probabilities)) {
            case 0: {
                timer->start("Add Fragment");
                bool success = networkMutator->addStartInstance(false);
                timer->stop("Add Fragment");
                if (success) {
                    return;
                } else {
                    reject();
                    probabilities[0] = 0;
                }
                break;
            }
            case 1: {
                timer->start("Remove Fragment");
                bool success = networkMutator->changeRandomInstance(true);
                timer->stop("Remove Fragment");
                if (success) {
                    return;
                } else {
                    reject();
                    probabilities[1] = 0;
                }
                break;
            }
            case 2: {
                timer->start("Modify Fragment");
                bool success = networkMutator->changeRandomInstance(false);
                timer->stop("Modify Fragment");
                if (success) {
                    return;
                } else {
                    reject();
                    probabilities[2] = 0;
                }
                break;
            }
        }

        if (globalSettings["Fewer Start Transitions"].get<bool>() && model->numSteps > 1) {
            return;
        }

        /*if (nodeStats.getCostChange("reject") > 0) {
            return;
        }*/

        done = true;
        for (double probability : probabilities) {
            if (probability > 0) {
                done = false;
                break;
            }
        }
    }
}

void GuideMutator::accept() {
    model->accept();
}

void GuideMutator::reject() {
    model->reject();
    /*nodeStats.restore();
    if (closeInspection(model->numSteps) || globalSettings.getBool("Debug Alerts")) {
        optimizer.verifyCost();
    }
    if (closeInspection(model->numSteps)) {
        std::cout << model->numSteps << " rejected." << std::endl;
    }*/
}

//void GuideMutator::save(Model* model) {
//    auto cost = optimizer.computeCost();
//    optimizer.accept(cost);
//    nodeStats.save();
//}

//bool GuideMutator::initializeModel(Model* model) {
//    model->setNodeStats(&nodeStats);
//    auto boundaryGroups = classifier->getBoundaryGroups();
//    
//    if (!boundaryGroups.empty()) {
//        auto emptyRing = classifier->getEmptyRing();
//        auto ringInstance = std::make_unique<RingInstance>(emptyRing, &nodeStats);
//        auto ringGroup = RingGroup::createFromRing(ringInstance->getRing());
//        auto startInstance = std::make_unique<RingGroupInstance>(ringGroup);
//        startInstance->addRingInstance(0, ringInstance.get());
//        
//        Transition transition{
//            startInstance.get(),
//            Util::pick(boundaryGroups),
//            true  // initialBoundary
//        };
//        
//        return hierarchyMutator->applyTransition(transition);
//    }
//    
//    return true;
//}

//void GuideMutator::reset() {
//    status = Status::UNRESOLVED;
//}
//
//void GuideMutator::pause() {
//    displayStats();
//    notify(true);
//    status = Status::PAUSED;
//}
//
//void GuideMutator::unresolve() {
//    status = Status::UNRESOLVED;
//}
//
//void GuideMutator::step(int numTasks) {
//    displayStats();
//    pauseCount = model->numSteps + numTasks;
//}

//void GuideMutator::setBreakTime(double newBreakTime) {
//    breakTime = newBreakTime;
//}
//
//void GuideMutator::setEndTime(double newEarlyEndTime, double newEndTime) {
//    earlyEndTime = newEarlyEndTime;
//    endTime = newEndTime;
//}

//Status GuideMutator::getStatus() const {
//    return status;
//}

//void GuideMutator::displayStats() {
//    std::string stats = "Iteration: " + std::to_string(model->numSteps) + " / " + 
//                       std::to_string(globalSettings.getInt("Max Iterations"));
//    
//    if (!globalSettings.getBool("MVP")) {
//        stats += " " + std::to_string(optimizer.getPrevCost().sum);
//    }
//    
//    // Update statistics display (implementation depends on UI system)
//    // document.getElementById('statistics').innerHTML = stats;
//}
//
//std::string GuideMutator::report() const {
//    auto cost = optimizer.getPrevCost();
//    std::string summary = std::to_string(cost.sum) + " " + 
//                         std::to_string(timer->getTiming("everything") / 1000.0) + "s " +
//                         std::to_string(model->numSteps / timer->getTiming("everything"));
//    
//    std::string details = cost.toString() + "\n" + timer->reportMean();
//    std::cout << "# of vertices: " << nodeStats.getCount("vertex") << std::endl;
//    
//    return summary + "\n" + details;
//}

//NodeStats* GuideMutator::getNodeStats() {
//    return &nodeStats;
//}
//
//Classifier* GuideMutator::getClassifier() const {
//    return classifier;
//}
//
//bool GuideMutator::closeInspection(int taskCount) {
//    if (forceContinued) {
//        return false;
//    }
//    
//    int taskStop = globalSettings.getInt("Task Stop");
//    int taskStep = globalSettings.getInt("Task Step");
//    
//    if (taskStop >= 0) {
//        return model->numSteps >= taskStop && ((model->numSteps - taskStop) % taskStep == 0);
//    }
//    
//    return globalSettings.getBool("Debug Each Task");
//}
//
//void GuideMutator::forceContinue() {
//    forceContinued = true;
//}
//
//void GuideMutator::hideStats() {
//    // Clear statistics display (implementation depends on UI system)
//    // document.getElementById('statistics').innerHTML = '';
//}
//
//Vertex* GuideMutator::findMutableVertex() {
//    auto vertexNodes = nodeStats.getNodes("vertex");
//    if (!vertexNodes.empty()) {
//        for (int i = 0; i < findMutableVertexAttempts; i++) {
//            if (auto vertex = dynamic_cast<Vertex*>(Util::pick(vertexNodes)->getElement())) {
//                if (vertex->isMoveable()) {
//                    return vertex;
//                }
//            }
//        }
//    }
//    return nullptr;
//}
//
//void GuideMutator::initializeHierarchyMutator() {
//    if (globalSettings.getBool("Use Network")) {
//        hierarchyMutator = std::make_unique<NetworkMutator>(classifier, &nodeStats);
//    } else {
//        hierarchyMutator = std::make_unique<FamilyTreeMutator>(classifier, &nodeStats);
//    }
//}

} // namespace ms 